### PR TITLE
Allow user to set all env variables

### PR DESCRIPTION
**Pull request by ayan-b**
Related to https://github.com/moremoban/moban/issues/202

Before raising the PR, here is a check list:

1) have you written unit tests for your code changes?
2) have you run "make format"?
3) are you requesting to "dev"?
4) have you updated the change log?
5) do you think that you can understand your changes after 6 month?
   5.1) can someone else understand your changes without your explanation?
6) are you pround of your code changes?
   6.1) do you have the feeling of achievement?
7) please add your name and github link to contributors.rst in alphabetical order.

### CODE REVIEW

**Comment by ayan-b** [comment]
@chfw what do you think of this approach? If it is acceptable, I will make tests and change the existing ones. I basically maintain a map `templatewise_env_vars` which maps `template_file_name` to `custom_env_vars`.

File: moban/jinja2/engine.py [hunk 1]
```diff
@@ -100,6 +100,9 @@ def get_template(self, template_file):
         template file exists at:
             '/User/moban-pro/my-template/templates/myfile.jj2'
         """
+        if custom_env_vars:
```
Comment (chfw) [inline moban/jinja2/engine.py]: custom env vars could be passed via jinja2's global plugin.

Comment (chfw) [reply moban/jinja2/engine.py]: https://github.com/moremoban/moban/blob/dev/moban/jinja2/extensions.py#L31

**Comment by chfw** [comment]
I thought `options` should be annotating the behaviour of the templating options.

as a side thought, passing global variables via `options` could be cheaper than jinja_globals as the latter require some programming knowledge.

File: moban/jinja2/engine.py (hunk 1 above)
Comment (jayvdb) [reply moban/jinja2/engine.py]: I think a concern is that each target could have different options.

tbis needs to nave tests written *before* it is reviewable. Tests communicate the goalsof a PR.

**Comment by chfw** [comment]
I suppose this PR shall resolve the following issue:

```
    options:
      ..
      block_end_string: '*))'
```

and regarding custom variables for each template action, I would suggest user to have their own data file, which then could choose to override whichever data file that the project has.

if this `options` becomes a short hand for passing custom variables, then it should be named as "options". I am not objecting this short-hand to pass custom variables.

**Comment by chfw** [comment]
please put this on hold. I am going to refactor (template_file, data_file, output) tuple so that it will be more flexible.

**Comment by ayan-b** [comment]
@chfw #202 aims to add the ability to modify all the custom env variables (block_start_string, block_end_string, variable_start_string, variable_end_string etc.).

**Comment by chfw** [comment]
but this could be expensive, as current settings one env per moban session. now this will introduce one env per template file.

**Comment by ayan-b** [comment]
> but this could be expensive, as current settings one env per moban session. now this will introduce one env per template file.

Yes, that's why I am updating the env vars inside `get_template` since the environment vars can be changed even after initialization of jinja2 environment.

**Comment by chfw** [comment]
however, since there is only one instance of env, once you updated it, you will need to make it un-set.

**Comment by ayan-b** [comment]
> however, since there is only one instance of env, once you updated it, you will need to make it un-set.

Yes, that is also true.

**Comment by chfw** [comment]
Let me think about it more.

**Comment by ayan-b** [comment]
> Let me think about it more.

@chfw what do you think? We can add some optimization like not resetting the `env` if it is not changed etc.

**Comment by chfw** [comment]
Let's close this PR till you will be ready

**Pull request closed without merging**